core/src/main/scala/akka/persistence/jdbc/journal/dao/JournalDao.scala

```diff
@@ -5,15 +5,12 @@
 
 package akka.persistence.jdbc.journal.dao
 
-import akka.NotUsed
-import akka.persistence.{ AtomicWrite, PersistentRepr }
-import akka.stream.scaladsl._
-
+import akka.persistence.AtomicWrite
 import scala.collection.immutable.Seq
 import scala.concurrent.Future
 import scala.util.Try
 
-trait JournalDao {
+trait JournalDao extends JournalDaoWithReadMessages {
```

Comment:
All that we document is https://doc.akka.io/docs/akka-persistence-jdbc/3.5/#custom-dao-implementation

Am i correct that this change is source compatible, but not binary compatible? In case of custom dao implementations? Maybe we should add a compatibility note for users who do implement a custom dao.

Reply:
Ah, the messagesWithBatch is implemented in `BaseJournalDaoWithReadMessages ` but not in `JournalDaoWithReadMessages`.

Would it perhaps be an option to do the "messagesWithBatch" magic in the journals itself. Like what we were doing in the `JdbcReadJournal`, or does that make it much harder to re-use the same code?

Reply:
My first comment here was about source compatibility.
Trying to guarantee binary compatibility for all these traits that are all public is pretty much impossible in my opinion.

If we think the compatibility is a problem I could move it to a standalone internal utility function that is used from both JdbcAsyncWriteJournal and JdbcReadJournal. It's not going to be very nice because there is no common interface for the `messages`. It's defined in both JournalDao and ReadJournalDao, so the `messages` must be wrapped by another function passed in to the utility.

Reply:
In the link you provided it says: "As these APIs are only now exposed for public use, the interfaces may change when the API needs to change for whatever reason."

Reply:
I don't like the idea of introducing a utility class. The implementation should be in the DAO (btw, it's was not the case before as part of it was in the JdbcReadJournal). 

I would prefer to keep binary compatibility, but if we see how this is being done today it's already very fragile. The user can provide a DAO impl, but it needs to have exactly the same constructor as the default DAO since we build it using reflection. 

We can also argue that this will be for `4.0.0` and that's not binary compatible.

Reply:
I am fine with anything as long as we make a note of this in our release notes